### PR TITLE
Write note witness cache atomically to disk to avoid corruption

### LINKED ISSUE #1378: error on startup: Assertion `nWitnessCacheSize >= nd->witnesses.size()' failed.
```
UpdateTip: new best=00311e4f2dae18b682086110e4ae2581857a0bb31e9eb66c127223aebffb9638  height=621  log2_work=17.079724  tx=632  date=2016-09-10 11:44:35 progress=1.000000  cache=0.2MiB(535tx)
AddToWallet 09d48c0a92b3a993ccc538555afe23388edc4c4743e473cd290a5bc6e0f6cafc  
UpdateTip: new best=0064a254e4bc19fbae2d8ef31761cfe3c2c529cd18b24d6578979e395069ba26  height=622  log2_work=17.085939  tx=634  date=2016-09-10 11:47:57 progress=1.000000  cache=0.2MiB(536tx)
AddToWallet 128e13f592d42b014181f92b8ba3ac395bd50238071a78c59dc6eb733b179f89  
zcashd: wallet/wallet.cpp:605: void CWallet::IncrementNoteWitnesses(const CBlockIndex*, const CBlock*, ZCIncrementalMerkleTree): Assertion `nWitnessCacheSize >= nd->witnesses.size()' failed.
Aborted
```

This is after my VM crashed and restarted; don't know whether that's relevant.

### PR DESCRIPTION
**Pull request by str4d**
Closes #1378

### CODE REVIEW

File: src/wallet/wallet.cpp [hunk 1]
```diff
@@ -679,13 +675,31 @@ void CWallet::DecrementNoteWitnesses()
...
-            walletdb.WriteWitnessCacheSize(nWitnessCacheSize);
+            WriteWitnessCache();
+        }
+    }
+}
+
+void CWallet::WriteWitnessCache() {
+    CWalletDB walletdb(strWalletFile);
+    if (!walletdb->TxnBegin()) {
+        // This needs to be done atomically, so don't do it at all
```
Comment (daira) [inline src/wallet/wallet.cpp]: And similarly for the three other abnormal returns below.

**Comment by daira** [comment]
Concept ACK. Log the errors as commented, and this needs a test.

**Comment by ebfull** [comment]
This is just a simple refactoring of existing code, which we decided we could not practically test (given our timeframe) in a previous PR. This is a race condition, we don't have the testing infrastructure to simulate something like this right now.

I'm ACK on this once the logs daira mentions are added.

**Comment by str4d** [comment]
Logging added, so putting the test aside per @ebfull's comment, this has two ACKs.

@zkbot r+

**Comment by str4d** [comment]
Rebased on master, fixed several bugs that escaped review, and squashed the commits. Please re-ACK.

**Comment by ebfull** [comment]
ACK again.

@zkbot r+

**Comment by daira** [comment]
What happens if an exception is thrown after ``TxnBegin()`` in ``WriteWitnessCache``? It should call ``TxnAbort`` and rethrow, I think.

**Pull request merged**